File: app/src/main/java/com/mgaetan89/showsrage/presenter/SchedulePresenter.java
package com.mgaetan89.showsrage.presenter;

import android.content.Context;
import android.support.annotation.Nullable;
import android.text.format.DateUtils;

import com.mgaetan89.showsrage.helper.DateTimeHelper;
import com.mgaetan89.showsrage.model.Indexer;
import com.mgaetan89.showsrage.model.Schedule;
import com.mgaetan89.showsrage.network.SickRageApi;

public class SchedulePresenter {
	@Nullable
	private Context context;

	@Nullable
	private Schedule schedule;

	public SchedulePresenter(@Nullable Schedule schedule, @Nullable Context context) {
		this.context = context;
		this.schedule = schedule;
	}

	@Nullable
	public CharSequence getAirDate() {
		if (this.schedule == null) {
			return null;
		}

		String airDate = this.schedule.getAirDate();

		if (airDate == null || airDate.isEmpty()) {
			return null;
		}

		return DateTimeHelper.getRelativeDate(airDate, "yyyy-MM-dd", DateUtils.DAY_IN_MILLIS);
	}

	@Nullable
	public CharSequence getAirDateTime() {
		CharSequence airDate = this.getAirDate();
		CharSequence airTime = this.getAirTime();

		if (airDate == null && airTime == null) {
			return null;
		}

		if (airDate != null) {
			if (airTime != null) {
				return airDate + ", " + airTime;
			}

			return airDate;
		}

		return airTime;
	}

	@Nullable
	public CharSequence getAirTime() {
		if (this.context == null || this.schedule == null) {
			return null;
		}

		String airDate = this.schedule.getAirDate();
		String airTime = this.getAirTimeOnly();

		if (airDate == null || airDate.isEmpty() || airTime == null || airTime.isEmpty()) {
			return null;
		}

		return DateTimeHelper.getLocalizedTime(this.context, airDate + " " + airTime, "yyyy-MM-dd K:mm a");
	}

	public int getEpisode() {
		if (this.schedule == null) {
			return 0;
		}

		return this.schedule.getEpisode();
	}

	public String getNetwork() {
		if (this.schedule == null) {
			return "";
		}

		return this.schedule.getNetwork();
	}

	public String getPosterUrl() {
		if (this.schedule == null) {
			return "";
		}

		return SickRageApi.getInstance().getPosterUrl(this.schedule.getTvDbId(), Indexer.TVDB);
	}

	public String getQuality() {
		if (this.schedule == null) {
			return "";
		}

		return this.schedule.getQuality();
	}

	public int getSeason() {
		if (this.schedule == null) {
			return 0;
		}

		return this.schedule.getSeason();
	}

	public String getShowName() {
		if (this.schedule == null) {
			return "";
		}

		return this.schedule.getShowName();
	}

	@Nullable
	/* package */ String getAirTimeOnly() {
		if (this.schedule == null) {
			return null;
		}

		String airTime = this.schedule.getAirs();

		if (airTime == null || airTime.isEmpty()) {
			return null;
		}

		return airTime.replaceFirst("(?i)^(monday|tuesday|wednesday|thursday|friday|saturday|sunday) ", "");
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/presenter/SchedulePresenter.kt
package com.mgaetan89.showsrage.presenter

import android.content.Context
import android.text.format.DateUtils
import com.mgaetan89.showsrage.helper.DateTimeHelper
import com.mgaetan89.showsrage.model.Indexer
import com.mgaetan89.showsrage.model.Schedule
import com.mgaetan89.showsrage.network.SickRageApi

class SchedulePresenter(val schedule: Schedule?, val context: Context?) {
	fun getAirDate(): CharSequence? {
		val schedule = this._getSchedule() ?: return null
		val airDate = schedule.airDate

		if (airDate.isEmpty()) {
			return null
		}

		return DateTimeHelper.getRelativeDate(airDate, "yyyy-MM-dd", DateUtils.DAY_IN_MILLIS)
	}

	fun getAirDateTime(): CharSequence? {
		val airDate = this.getAirDate()
		val airTime = this.getAirTime()

		if (airDate == null && airTime == null) {
			return null
		}

		if (airDate != null) {
			return if (airTime != null) "$airDate, $airTime" else airDate
		}

		return airTime
	}

	private fun getAirTime(): CharSequence? {
		val schedule = this._getSchedule() ?: return null

		if (this.context == null) {
			return null
		}

		val airDate = schedule.airDate
		val airTime = this.getAirTimeOnly()

		if (airDate.isEmpty() || airTime.isNullOrEmpty()) {
			return null
		}

		return DateTimeHelper.getLocalizedTime(this.context, "$airDate $airTime", "yyyy-MM-dd K:mm a")
	}

	fun getEpisode() = this._getSchedule()?.episode ?: 0

	fun getNetwork() = this._getSchedule()?.network ?: ""

	fun getPosterUrl(): String {
		val schedule = this._getSchedule() ?: return ""

		return SickRageApi.instance.getPosterUrl(schedule.tvDbId, Indexer.TVDB)
	}

	fun getQuality() = this._getSchedule()?.quality ?: ""

	fun getSeason() = this._getSchedule()?.season ?: 0

	fun getShowName() = this._getSchedule()?.showName ?: ""

	internal fun getAirTimeOnly(): String? {
		val schedule = this._getSchedule() ?: return null
		val airTime = schedule.airs

		if (airTime.isEmpty()) {
			return null
		}

		return airTime.replaceFirst("(?i)^(monday|tuesday|wednesday|thursday|friday|saturday|sunday) ".toRegex(), "")
	}

	internal fun _getSchedule() = if (this.schedule?.isValid == true) this.schedule else null
}
